src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -907,14 +923,122 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
             IPartition<TSource> partition = source as IPartition<TSource>;
             if (partition != null) return partition.Skip(count);
             IList<TSource> sourceList = source as IList<TSource>;
-            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+            return sourceList != null ? new SkipListIterator<TSource>(sourceList, count, int.MaxValue) : SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        private sealed class SkipListIterator<TSource> : Iterator<TSource>, IPartition<TSource>
         {
-            while (count < source.Count)
+            private readonly IList<TSource> _source;
+            private readonly int _minIndex;
+            private readonly int _maxIndex;
+            private int _index;
+
+            public SkipListIterator(IList<TSource> source, int minIndex, int maxIndex)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(minIndex >= 0);
+                Debug.Assert(minIndex <= maxIndex);
+                _source = source;
+                _minIndex = minIndex;
+                _index = minIndex;
+                _maxIndex = maxIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new SkipListIterator<TSource>(_source, _minIndex, _maxIndex);
+            }
+
+            public override bool MoveNext()
+            {
+                if ((state == 1 & _index <= _maxIndex) && _index < _source.Count)
+                {
+                    current = _source[_index];
+                    ++_index;
+                    return true;
+                }
+                Dispose();
+                return false;
+            }
+
+            public IPartition<TSource> Skip(int count)
+            {
+                int minIndex = _minIndex + count;
+                return minIndex >= _maxIndex
+                    ? (IPartition<TSource>)new EmptyPartition<TSource>()
+                    : new SkipListIterator<TSource>(_source, minIndex, _maxIndex);
+            }
+
+            public IPartition<TSource> Take(int count)
+            {
+                int maxIndex = _minIndex + count - 1;
+                return new SkipListIterator<TSource>(_source, _minIndex, maxIndex >= _maxIndex ? _maxIndex : maxIndex);
```

Comment:
Yes. I very much under-use that and `Max` generally, and this is a case where I really should have.